src/vs/editor/browser/services/treeSitter/treeSitterParserService.ts

```diff
@@ -80,13 +84,12 @@ class TextModelTreeSitter extends Disposable {
 			}
 		}
 
-		this._treeSitterTree.tree = await this._treeSitterParser.parse(this.model, this._treeSitterTree);
+		treeSitterTree.tree = await this._treeSitterParser.parse(this.model, treeSitterTree);
```

Comment:
I think this could still be problematic, if the first parse call takes longer than the second parse call.
Probably this cannot happen, but I think it is worth to write the code in a way that there is no doubt about it.
I think this mutation should be done by a method in `treeSitterTree`, which maybe maintains a queue of parse requests or something to serialize the updates.

Reply:
Added a queue in `TextModelTreeSitter`.

Reply:
Per our discussion, I've moved the queue into `TreeSitterTree`.